### ISSUE TITLE
AttentionWrapper zero_state(batch_size, tf.float32).clone(cell_state=encoder_state) fails when batch size is 1

### ISSUE THREAD

**Issue by nikita68** [tensorflow/tensorflow#15737]
Hello!
I believe to have found a small bug when using the `zero_state(batch_size, tf.float32).clone(cell_state=encoder_state)` command. When batch size is 1, the error `ValueError: The shape for decoder/while/Merge_5:0 is not an invariant for the loop. It enters the loop with shape (1, 512), but has shape (?, 512) after one iteration. Provide shape invariants using either the shape_invariants argument of tf.while_loop or set_shape() on the loop variables.` is thrown. This error does not occur when batch size is 2 or larger. The error also doesn't occur if I remove the .clone command. 
I tried investigating where the error is, but couldn't find the cause. I'm using this in context of trying to build a neural transducer, but also get the same error for basic seq2seq:
(Based on the NMT tutorial)

```python
# .... Encoder, constants etc...
# Decoder
helper = tf.contrib.seq2seq.TrainingHelper(
    decoder_inputs_embedded, decoder_full_length, time_major=True) 

attention_states = tf.transpose(encoder_outputs, [1, 0, 2])  # attention_states: [batch_size, max_time, num_units]
attention_mechanism = tf.contrib.seq2seq.LuongAttention(
    encoder_hidden_units, attention_states,
    memory_sequence_length=encoder_inputs_length)
decoder_cell = tf.contrib.seq2seq.AttentionWrapper(
    tf.contrib.rnn.LSTMCell(decoder_hidden_units),
    attention_mechanism,
    attention_layer_size=decoder_hidden_units)

projection_layer = layers_core.Dense(
    vocab_size, use_bias=False)


decoder = tf.contrib.seq2seq.BasicDecoder(
    decoder_cell, helper, decoder_cell.zero_state(batch_size, tf.float32).clone(cell_state=encoder_state),
    output_layer=projection_layer)

# ---- Training ----
outputs, last_state, _ = tf.contrib.seq2seq.dynamic_decode(decoder, output_time_major=True)
logits = outputs.rnn_output
decoder_prediction = outputs.sample_id
```

TF Version: ('v1.4.0-19-ga52c8d9', '1.4.1')
System details:
```
== cat /etc/issue ===============================================
Linux nikita-coolboi 4.13.0-21-generic #24-Ubuntu SMP Mon Dec 18 17:29:16 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
VERSION="17.10 (Artful Aardvark)"
VERSION_ID="17.10"
VERSION_CODENAME=artful

== are we in docker =============================================
No

== compiler =====================================================
c++ (Ubuntu 7.2.0-8ubuntu3) 7.2.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


== uname -a =====================================================
Linux nikita-coolboi 4.13.0-21-generic #24-Ubuntu SMP Mon Dec 18 17:29:16 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

== check pips ===================================================
numpy (1.13.3)
protobuf (3.5.1)
tensorflow (1.4.1)
tensorflow-tensorboard (0.4.0rc3)

== check for virtualenv =========================================
False

== tensorflow import ============================================
tf.VERSION = 1.4.1
tf.GIT_VERSION = v1.4.0-19-ga52c8d9
tf.COMPILER_VERSION = v1.4.0-19-ga52c8d9
Sanity check: array([1], dtype=int32)

== env ==========================================================
LD_LIBRARY_PATH is unset
DYLD_LIBRARY_PATH is unset

== nvidia-smi ===================================================
tf.sh: line 105: nvidia-smi: command not found

== cuda libs  ===================================================
```
I believe this is an important issue, as often times when experimenting with new seq2seq models I start off with trying to get it to work for a batch size of 1.

Thanks!
Nikita

**Comment by bignamehyp** [comment]
@ebrevdo can you please take a look. Thanks.

**Comment by ebrevdo** [comment]
This looks like an important bug to fix.  Can you provide a pre-amble creating properly sized input tensors, like `decoder_inputs_embedded`, `encoder_outputs`, etc, using `tf.ones` or `tf.random_...`, so that I can directly replicate the issue?

**Comment by ebrevdo** [comment]
CC += @oahziur

**Comment by nikita68** [comment]
@ebrevdo 
I've made two very hacky standalone code samples, and I found out, that this error only occurs when using placeholders:

This example fails only when the clone method is called:
```python
import tensorflow as tf
import numpy as np
from tensorflow.python.layers import core as layers_core

# -- Constants ---
vocab_size = 10
input_dimensions = 20
input_embedding_size = 50
encoder_hidden_units = 512
decoder_hidden_units = encoder_hidden_units
batch_size = 1
time_to_sim_input = 5

# ---- Build model ----
encoder_inputs = tf.placeholder(shape=(None, None, input_dimensions), dtype=tf.float32, name='encoder_inputs')
encoder_inputs_length = tf.placeholder(shape=(None,), dtype=tf.int32, name='encoder_inputs_length')
decoder_targets = tf.placeholder(shape=(None, None), dtype=tf.int32, name='decoder_targets')
decoder_target_length = tf.placeholder(shape=(None,), dtype=tf.int32, name='decoder_target_length')
decoder_inputs = tf.placeholder(shape=(None, None), dtype=tf.int32, name='decoder_inputs')
decoder_full_length = tf.placeholder(shape=(None,), dtype=tf.int32, name='decoder_full_length')


embeddings = tf.Variable(tf.random_uniform([vocab_size, input_embedding_size], -1.0, 1.0), dtype=tf.float32)
encoder_inputs_embedded = encoder_inputs
decoder_inputs_embedded = tf.nn.embedding_lookup(embeddings, decoder_inputs)
decoder_inputs_embedded = tf.transpose(decoder_inputs_embedded, perm=[1, 0, 2])  # Make it time major again


# ---- Encoder ------
encoder_cell = tf.contrib.rnn.LSTMCell(encoder_hidden_units)

# Run Dynamic RNN
#   encoder_outputs: [max_time, batch_size, num_units]
#   encoder_state: [batch_size, num_units]
encoder_outputs, encoder_state = tf.nn.dynamic_rnn(
    encoder_cell, encoder_inputs_embedded,
    sequence_length=encoder_inputs_length, time_major=True,
    dtype=tf.float32)


# ---- Decoder -----
helper = tf.contrib.seq2seq.TrainingHelper(
    decoder_inputs_embedded, decoder_full_length, time_major=True)

attention_states = tf.transpose(encoder_outputs, [1, 0, 2])  # attention_states: [batch_size, max_time, num_units]
attention_mechanism = tf.contrib.seq2seq.LuongAttention(
    encoder_hidden_units, attention_states,
    memory_sequence_length=encoder_inputs_length)
decoder_cell = tf.contrib.seq2seq.AttentionWrapper(
    tf.contrib.rnn.LSTMCell(decoder_hidden_units),
    attention_mechanism,
    attention_layer_size=decoder_hidden_units)

projection_layer = layers_core.Dense(
    vocab_size, use_bias=False)


# ------------------------- CLONE --------------------------
decoder = tf.contrib.seq2seq.BasicDecoder(
    decoder_cell, helper, decoder_cell.zero_state(batch_size, tf.float32).clone(cell_state=encoder_state),
    output_layer=projection_layer)

# ---- Training ----
outputs, last_state, _ = tf.contrib.seq2seq.dynamic_decode(decoder, output_time_major=True)
logits = outputs.rnn_output
decoder_prediction = outputs.sample_id


targets_one_hot = tf.one_hot(decoder_targets, depth=vocab_size, dtype=tf.float32)
stepwise_cross_entropy = tf.nn.softmax_cross_entropy_with_logits(labels=targets_one_hot, logits=logits)
loss = tf.reduce_mean(stepwise_cross_entropy)
train_op = tf.train.AdamOptimizer().minimize(loss)

init = tf.global_variables_initializer()


# ---- Loader -----


def next_batch(amount=batch_size):

    e_in = np.random.uniform(-1.0, 1.0, size=(batch_size, time_to_sim_input, input_dimensions))
    e_in_length = [time_to_sim_input]
    d_targets = np.asarray([[1, 2, 3, 4, 5]])
    d_targets_length = [len(np.asarray([1, 2, 3, 4, 5]))]
    offset_din = np.asarray([[9, 1, 2, 3, 4, 5]])

    return {
        encoder_inputs: np.transpose(e_in, axes=[1, 0, 2]),
        encoder_inputs_length: e_in_length,
        decoder_targets: np.transpose(d_targets),
        decoder_target_length: d_targets_length,
        decoder_inputs: offset_din,
        decoder_full_length: np.asarray([len(np.asarray([1, 2, 3, 4, 5]))] * amount),
    }


with tf.Session() as sess:
    sess.run(init)
    losses = []

    feed = next_batch()
    _, l, predict = sess.run([train_op, loss, decoder_prediction], feed)
    losses.append(l)

    print 'Finished'

```

And this example is fine, but doesn't use placeholders:

```python
import tensorflow as tf
from tensorflow.python.layers import core as layers_core

# -- Constants ---
vocab_size = 10  # PAD = 8, EOS = 9
input_dimensions = 20
input_embedding_size = 50
encoder_hidden_units = 512
decoder_hidden_units = encoder_hidden_units
batch_size = 1
time_to_sim_input = 8

# ---- Build model ----

encoder_inputs = tf.random_uniform(shape=(time_to_sim_input, batch_size, input_dimensions))
encoder_inputs_length = tf.constant(time_to_sim_input, shape=(batch_size,))
decoder_targets = tf.constant([1, 2, 3, 4, 5], shape=(5, batch_size))
decoder_targets_length = tf.constant(4, shape=(batch_size,))
decoder_inputs = tf.constant([9, 1, 2, 3, 4], shape=(batch_size, 5))
decoder_full_length = tf.constant([5] * batch_size, shape=(batch_size,))

embeddings = tf.Variable(tf.random_uniform([vocab_size, input_embedding_size], -1.0, 1.0), dtype=tf.float32)
encoder_inputs_embedded = encoder_inputs
decoder_inputs_embedded = tf.nn.embedding_lookup(embeddings, decoder_inputs)
decoder_inputs_embedded = tf.transpose(decoder_inputs_embedded, perm=[1, 0, 2])  # Make it time major again


# ---- Encoder ------
encoder_cell = tf.contrib.rnn.LSTMCell(encoder_hidden_units)

# Run Dynamic RNN
#   encoder_outputs: [max_time, batch_size, num_units]
#   encoder_state: [batch_size, num_units]
encoder_outputs, encoder_state = tf.nn.dynamic_rnn(
    encoder_cell, encoder_inputs_embedded,
    sequence_length=encoder_inputs_length, time_major=True,
    dtype=tf.float32)

# ---- Decoder -----
helper = tf.contrib.seq2seq.TrainingHelper(
    decoder_inputs_embedded, decoder_full_length, time_major=True)

attention_states = tf.transpose(encoder_outputs, [1, 0, 2])  # attention_states: [batch_size, max_time, num_units]
attention_mechanism = tf.contrib.seq2seq.LuongAttention(
    encoder_hidden_units, attention_states,
    memory_sequence_length=encoder_inputs_length)
decoder_cell = tf.contrib.seq2seq.AttentionWrapper(
    tf.contrib.rnn.LSTMCell(decoder_hidden_units),
    attention_mechanism,
    attention_layer_size=decoder_hidden_units)

projection_layer = layers_core.Dense(
    vocab_size, use_bias=False)

# ------------------------- CLONE --------------------------
decoder = tf.contrib.seq2seq.BasicDecoder(
    decoder_cell, helper, decoder_cell.zero_state(batch_size, tf.float32).clone(cell_state=encoder_state),
    output_layer=projection_layer)

# ---- Training ----
outputs, last_state, _ = tf.contrib.seq2seq.dynamic_decode(decoder, output_time_major=True)
logits = outputs.rnn_output
decoder_prediction = outputs.sample_id


targets_one_hot = tf.one_hot(decoder_targets, depth=vocab_size, dtype=tf.float32)
stepwise_cross_entropy = tf.nn.softmax_cross_entropy_with_logits(labels=targets_one_hot, logits=logits)
loss = tf.reduce_mean(stepwise_cross_entropy)
train_op = tf.train.AdamOptimizer().minimize(loss)

init = tf.global_variables_initializer()


# ---- Loader -----

with tf.Session() as sess:
    sess.run(init)
    losses = []

    _, l, predict = sess.run([train_op, loss, decoder_prediction])
    losses.append(l)

    print 'Finished'

```

**Comment by oahziur** [comment]
@nikita68 

I think you need to make sure the the encoder_state has a static batch_size which means you need to set your placeholder's batch dimension as `batch_size` instead of `None`.

 or 

you can just use a tensor as the zero_states' batch size instead of a constant.

For example:

```
decoder_init_state = decoder_cell.zero_state(tf.size(decoder_full_length), tf.float32).clone(cell_state=encoder_state)
```

**Comment by ebrevdo** [comment]
Do we need to, in general, have a static batch size for the encoder?  We
shouldn't require that.

**Comment by oahziur** [comment]
@ebrevdo The encoder doesn't need to have static batch size, as long as we don't give a static vale to the zero_states in this case.

I think the problem here is clone() copies an encoder state without static batch size to an attention wrapper state that was initialized with static batch size.

**Comment by ebrevdo** [comment]
Ah!  I see; that's why it lost that size information.  I wonder if we can
bring that static shape info back in, inside the decode body.

**Comment by nikita68** [comment]
@oahziur Thanks, it fixed the problem.
@ebrevdo I've also been thinking the same thing.
Is this issue considered closed?

**Comment by ebrevdo** [comment]
@nikita68 probably not; @oahziur is looking at possible solutions.  we had one that made `clone()` perform some additional static shape inference, but that's a bit fraught.  probably we can add some `set_shape` calls inside either `AttentionWrapper`.  @oahziur wdyt?

**Comment by oahziur** [comment]
@ebrevdo 

`clone()` maybe the best place to set the new Tensor's batch_size info since that is the only place users may update tensors in the `AttentionWrapperState`.

Instead of doing `set_shape` to copy the static shape information, I think we can use the attention_mechanism.batch_size for the initialization in `AttentionWrapper.zero_state()` to fix issues like this, so the batch_size info is always consistent with `memory` and `encoder_state`.

**Comment by dyelax** [comment]
Was this resolved? I'm still having this issue in 1.4
